main/guides/js-programming/far.md

```diff
@@ -0,0 +1,73 @@
+# `Far()` and Remotable Objects
+
+## Remotable objects
+
+Some objects have methods that can be invoked remotely (across contracts and vats) using
```

Comment:
I think as written, this is very confusing for an outsider. I think we need to back up and (if necessary) dial down the precision in order to achieve something understandable. 

How about starting out with something like the following instead, corrected for any mistakes:

---
In Agoric smart contracts and dapps, you can call methods on objects that originate in another vat or on another machine. For example, a purse for an ERTP issuer actually lives in the vat in which the issuer was created, but you can use the purse in your off-chain wallet or in another contract. In other words, Agoric encapsulates all of the inter-machine and inter-vat communication so that at the level of a smart contract or dapp, objects from other vats or machines can be treated almost  as if they are local. 

In order to call a method on an object from another vat or other machine, you must use `E()`, such as `E(issuer).getBrand()`. 

Objects (like ERTP issuers) which are intended to be used in other vats are called `Remotables`. Importantly, any object returned from a smart contract, such an `offerResult`, `publicFacet` or `creatorFacet`, must be `Remotable`.

## Rules for Creating Remotables
1. All property names must be strings. Property names must not be Symbols.
3. All property values must be functions. (Note: If you wish to send data, send a pass-by-copy record instead, or add a function that returns a pass-by-copy record.)
4. The object must be wrapped with `Far`. (Note: `Far` also hardens the object.)

## Rules for Using Remotables
1. Call the method by first wrapping the object with `E`, such as `E(issuer).getBrand()`;
2. Handle the resulting promise. Calling `E()` always results in a promise.

## How to use Far to Correctly Wrap a Remotable
...

Reply:
I like @katelynsills ' version.

@katelynsills ' one thing seems off. An `offerResult` just needs to be passable. I don't see and reason it needs to be remotable. This is true for the `publicFacet` and `creatorFacet` as well, but less important since they are much more likely to be remotables in practice.

@tyg Do we have a glossary somewhere where we try to explain our terminology precisely? For a lot of what I explained on that github thread, I think @katelynsills is probably right to omit them here. But these terms should be explained precisely somewhere.

Reply:
> One thing seems off. An offerResult just needs to be passable. I don't see any reason it needs to be remotable. This is true for the publicFacet and creatorFacet as well, but less important since they are much more likely to be remotables in practice.

Oh, good point. That's right. The `offerResult` could be a pass-by-copy object or not an object at all.

Reply:
@erights We have the general glossary (visible at https://agoric.com/documentation/glossary/, in GitHub at documentation/main/glossary, but I'm not sure that syncs up with what you're looking for; it consists of relatively short definitions intended more to give users an initial idea of what we're talking about. Your sort of thing is more like the deep dives Allison's using in your Foresight book and are fairly detailed and precise discussions as much as definitions, IMO.

Reply:
Would a Purse, Payment, or Offer be a better example of a Remotable? In particular, are they the most-well-represented objects in our docs and sample code?

I think the `publicFacet` and `creatorFacet` will generally *not* be Remotable, since they'll have all sorts of random data as properties.

Reply:
> I think the publicFacet and creatorFacet will generally not be Remotable, since they'll have all sorts of random data as properties.

We don't have any publicFacets or creatorFacets that aren't Remotable. By design, they must be Remotable, but I don't believe we enforce this anywhere at the Zoe level. 

> Would a Purse, Payment, or Offer be a better example of a Remotable? 

A purse is a good example of using a Remotable, but not of creating your own Remotable object, since that's done for you in ERTP. A payment isn't a good example of either using or creating, because you don't usually call methods on a payment. An offer proposal is a pass-by-copy record. The UserSeat that you get back when you make an offer could be a good example of Remotable usage, but not creation, since the user doesn't make the object.

Reply:
I'm a little confused at this point. Since methods are called using payment as an argument, do payments need to be far()ed when used as arguments to a method on an Issuer object? And other than that, it seems that anything you get from an IssuerKit needs to be far()ed? But Amounts don't need to be far()ed, since they're now made locally via amountMath.make()? Also, if all the IssuerKit results will need far(), is there any reason not to just call far() on them internally just before returning them so the developer doesn't have to?

And is there any harm in far()ing an object that doesn't need to be? If so, is there any sort of error or warning that will catch doing so?

Reply:
ERTP does all of the `Far` calls that are necessary, so there's no need to wrap a payment with `Far` - issuer.js already does that. 

> anything you get from an IssuerKit needs to be far()ed?

Nope, already handled for you.

> But Amounts don't need to be far()ed, since they're now made locally via amountMath.make()

Amounts are records with no functions as values so they are never wrapped with `Far`

> Also, if all the IssuerKit results will need far(), is there any reason not to just call far() on them internally just before returning them so the developer doesn't have to?

IssuerKit results don't need `Far` - that's handled internally.

> And is there any harm in far()ing an object that doesn't need to be? If so, is there any sort of error or warning that will catch doing so?

There's no harm in wrapping an object with `Far` even if it never leaves the vat. There's no error or warning, as there's nothing wrong with this. Wrapping an object that is actually a record and doesn't have function values will error, I believe. I can check this.

Reply:
If an object, `foo`, should never be exposed to other vats, there is a benefit to not declaring it `Far`, i.e., not making it a Remotable. If `foo` is not marked but is accidentally exposed, that will cause an error, suppressing the vulnerability that might have followed from such accidental exposure.

We don't have this benefit until `Far` or `Remotable` are mandatory, which we're close to. Certainly, the documentation should assume that it is mandatory.